django/contrib/admin/paginations.py

```diff
@@ -0,0 +1,116 @@
+from django.contrib.admin.options import IncorrectLookupParameters
+from django.core.paginator import InvalidPage
+from django.template.defaulttags import querystring
+from django.utils.html import format_html, format_html_join
+from django.utils.safestring import mark_safe
+
+ALL_VAR = "all"
+PAGE_VAR = "p"
+
+
+class Pagination:
+    def __init__(
+        self,
+        request,
+        model,
+        list_per_page,
+        list_max_show_all,
+        queryset,
+        model_admin,
+    ):
+        self.model = model
+        self.opts = model._meta
+        self.list_per_page = list_per_page
+        self.list_max_show_all = list_max_show_all
+        self.model_admin = model_admin
+        try:
+            self.page_num = int(request.GET.get(PAGE_VAR, 1))
+        except ValueError:
+            self.page_num = 1
+        self.show_all = ALL_VAR in request.GET
+        self.filter_params = request.GET.copy()
+        self.queryset = queryset
+        self.setup(request)
+
+    @property
+    def pagination_required(self):
+        return (not self.show_all or not self.can_show_all) and self.multi_page
```

Comment:
```suggestion
        self.multi_page and not (self.show_all and self.can_show_all)
```
I think reversing the order here is more readable

Reply:
I agree. Thank you, cliff!